subsys/bluetooth/host/l2cap.c

```diff
@@ -1983,17 +1985,25 @@ static int l2cap_chan_le_send(struct bt_l2cap_le_chan *ch,
 
 	len = seg->len - sdu_hdr_len;
 
-	/* Set a callback if there is no data left in the buffer */
-	if (buf == seg || !buf->len) {
-		err = bt_l2cap_send_cb(ch->chan.conn, ch->tx.cid, seg,
-				       l2cap_chan_sdu_sent,
-				       l2cap_tx_meta_data(buf));
+	/* SDU will be considered sent when there is no data left in the
+	 * buffers, or if there will be no data left, if we are sending `buf`
+	 * directly.
+	 */
+	if (!net_buf_frags_len(buf) ||
+	    (buf == seg && net_buf_frags_len(buf) == len)) {
+		cb = l2cap_chan_sdu_sent;
 	} else {
-		err = bt_l2cap_send_cb(ch->chan.conn, ch->tx.cid, seg,
-				       l2cap_chan_seg_sent,
-				       l2cap_tx_meta_data(buf));
+		cb = l2cap_chan_seg_sent;
 	}
 
+	/* Forward the PDU to the lower layer.
+	 *
+	 * Note: after this call, anything in buf->user_data should be
+	 * considered lost, as the lower layers are free to re-use it as they
+	 * see fit. Reading from it later is obviously a no-no.
+	 */
+	err = bt_l2cap_send_cb(ch->chan.conn, ch->tx.cid, seg, cb, l2cap_tx_meta_data(buf));
+
```

Comment:
This refactor should ideally be a free-standing commit. The change is small but the diff is large, and it is mixed up with a significant change in the if's condition.